### DISCUSSION TITLE
How to install Composer so that global packages are in the PATH?

### DISCUSSION THREAD

**Discussion by dreadnaut** [composer/composer#12124] [Q&A / Support] [answered]
Hello folks 👋
@tomasr1981 and I are looking at their Composer setup, and [something seems odd](https://github.com/dreadnaut/PHPloy/issues/1#issuecomment-2363239238) in the configuration: global packages are not running, and even after adding the Composer's directory to the PATH with
```sh
export PATH=/home/user/.config/composer/vendor/bin:$PATH
```
things don't seem alright.

The original issue was to get `dreadnaut/phploy` running, but we can reproduce it with `squizlabs/php_codesniffer` (which I trust more than my own package).

@tomasr1981 can probably give you more details on how Composer was installed on his (Linux?) system, but the general question is: is there a suggested way of installing Composer so that global packages land in the right place, and can be executed without trouble?

(I'm on Windows, and I could only help up to certain point, so here we are!)

**Comment by Seldaek** [comment]
setting the PATH to include the global bin dir should be enough.. But that may not apply until you start a new shell (or explicitly run the `export PATH=...` line in the current shell). On windows env vars like PATH also need a new terminal to be started to receive the new env vars after they change.

**Reply by tomasr1981** [reply]
Hi,
I install global packages (the sample is for squizlabs/php_codesniffer, but I have the same problem with dreadnaut/phploy) via composer in this way, but calling from the project does not work either. It seems like it keeps taking a different route than it should.

```
~$ composer global require "squizlabs/php_codesniffer=*"
Changed current directory to /home/user/.config/composer
The package you required is recommended to be placed in require-dev (because it is tagged as "static analysis") but you did not use --dev.
Do you want to re-run the command with --dev? [yes]? y
./composer.json has been updated
Running composer update squizlabs/php_codesniffer
Loading composer repositories with package information
Updating dependencies
Lock file operations: 1 install, 0 updates, 0 removals
  - Locking squizlabs/php_codesniffer (3.10.3)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 1 install, 0 updates, 0 removals
  - Downloading squizlabs/php_codesniffer (3.10.3)
  - Installing squizlabs/php_codesniffer (3.10.3): Extracting archive
Generating autoload files
3 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
No security vulnerability advisories found.
```

```
/projects/project$ phpcs
The "phpcs" command was not found, but it can be installed using:
sudo apt install php-codesniffer

```
```
Export path does not help:
export PATH=/home/user/.config/composer/vendor/bin:$PATH
```

I use linux ubuntu (mint).
Can you tell me what I'm doing wrong?

**Reply by Seldaek** [reply]
can you do `ls /home/user/.config/composer/vendor/bin` and see what's in there?

try `hash -r` as well to reset bash's cache but if it didn't find anything I don't think it caches it so it probably doesn't help.

**Reply by tomasr1981** [reply]
First command returned `phpbf phpcs phploy`.
Second one did not help.

**Reply by xabbuh** [reply]
What's the output of `echo $PATH`?

**Reply by tomasr1981** [reply]
In root
```
~$ echo $PATH
/home/user/.config/composer/vendor/bin:~/.config/.composer/vendor/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin

```
In project
`~/projects/project$ echo $path
`

There is empty string. Is that the problem? Do I have to export path in every project?

**Reply by stof** [reply]
@tomasr1981 be careful. `echo $PATH` and `echo $path` are different things.

**Reply by tomasr1981** [reply]
Oh, I'm sorry.
```
~/projects/project$ echo $PATH
~/.config/.composer/vendor/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
```

**Reply by xabbuh** [reply]
There's one difference: `composer` vs. `.composer` (note the leading dot)

**Reply by stof** [reply, marked as answer]
The `~/.config/.composer/vendor/bin` path is wrong.

and your custom `export PATH` adding another (working) path is not available in the project because each terminal has its own env variables and the `export` command affects only the terminal process in which you run it. To make it permanent, you need to do it in your bash config file instead (which is probably where the broken `~/.config/.composer/vendor/bin` is defined so you could fix it)

**Reply by tomasr1981** [reply]
So I found the `.profile` file where this path was. I removed the dot from `.composer` and after a reboot everything works. Thank you all for your help.

**Comment by stof** [comment]
2 things to note if your are on Windows:

- on Windows, the PATH_SEPARATOR is `;`, not `:` (as `:` appears as part of absolute file paths)
- `/home/user/.config/composer/vendor/bin` will not be the path to your global bin directory on Windows (this does not look like a Windows path). Use `composer global config bin-dir --absolute` to make Composer output the absolute path to this folder.

**Comment by dreadnaut** [comment]
Apologies for the confusion, let me clarify:

- I'm on Windows, and have no trouble: Composer's installer is setting up directories and PATH for me
- I'm trying to help @tomasr1981 solve their problem, and he's on a Linux system (as far as I understand); all paths and comments in my opening post are relative to his context

Thank you for your help!
